designs/service_function_chaining/SFC_design.md

```diff
@@ -0,0 +1,127 @@
+
+A basic example of IP packet forwarding, with respect to SFC is, packets that are from point A to point B will be redirected to an IPS or Firewall VMs (Service Functions) , even though the service functions are not in between their route.  
+
+The SF’s can be chained to make packets go through different Service Appliances(IPS,FIREWALL.  
+
+### SFC Concepts
+- Service Functions (SF) . Like IPS, firewalls ..etc.  
+
+- Port-Pairs : A port pair represents a service function instance that includes an ingress and egress port.  
+- Port-Pair-Groups : A port pair group may contain one or more port pairs. Multiple port pairs enable load balancing/distribution over a set of functionally equivalent service functions.  
+- Port-Chain : A port chain consists of a sequence of port pair groups. Each port pair group is a hop in the port chain. A group of port pairs represents service functions providing equivalent functionality. A flow classifier attached to port-chain will steer the traffic into the port-chain.  
+- Flow-Classification : A flow classifier identifies a flow. A port chain can contain multiple flow classifiers. Omitting the flow classifier effectively prevents steering of traffic through the port chain.  
+
+### SFC Configuration
+- SFC is neutron extension in Openstack compute node
+- OVS will be used as backend driver
+- Current configuration on Ocata red hat Openstack platform.
+
+SFC work flow in Openstack Environment.
+![](./images/sfc-dataflow.png)
+
+
+## Design Changes  
+SFC functionality in OSC will adapt to existing SDN controller SDK APIs.SFC configuration work flow, changes to SDK APIs and interfaces to support SFC plugin are described below in detail. Other supported SDN controller plugins should adapt to all the below mention changes.
+
+![](./images/sw-modules.png)
+### REST API   
+Currently not applicable.
+ 
+### OSC SDKs  
+
+#### SDN Controller SDK  
+Following section describes in detail changes in SDN controller SDK APIs and interfaces to support SFC plugin.
+
+
+##### Inspection port
+OSC core will create inspection port during creation of new distribution specification (DS) under a distribution Appliance.  
+
+SDN controller SDK APIs `registerinspectionport` and `removeinspectionport` will be called to create and remove the inspection port with `InspectionPortElement` interface as the argument.  
+
+SFC plugin should adapt the same SDK controller APIs to create and delete inspection port hook entity in OSC. 
+
+Following changes are proposed with respect to Inspection port actions.
+
+`InspectionPortElement` interface will have an additional field `redirection-target-id`. For more understanding of `redirection-target-id` follow inspection port work flow described below.

```

Comment:
we should introduce a new interface for the return type which can have only redirection-target-id. (i would call this something more generic) 
@larkinscarvalho can explain rationale

we can still add redirection-target-id to the InspectionPortElement interface since this is needed when registering the ports.

Also we can split the Network Element interface into 2 and extract the getElementId and getParentId methods to a new interface like "Element" while making the network element interface inherit this "Element" interface.

I would look at http://www.oracle.com/technetwork/java/codeconventions-135099.html for the method/variables and interface names.